### DISCUSSION TITLE
Lesson 12 Test MintDsc

### DISCUSSION THREAD

**Discussion by 0x-codeX** [Cyfrin/foundry-full-course-cu#1611] [Q&A]
I get the same error message when I try to test the two below function under MintDsc.

`
function testRevertsIfMintFails() public {
        // Arrange - Setup
        //MockFailedMintDSC mockDsc = new MockFailedMintDSC();
        tokenAddresses = [weth];
        priceFeedAddresses = [ethUsdPriceFeed];
        address owner = msg.sender;
        vm.prank(owner);
        DSCEngine mockDsce = new DSCEngine(tokenAddresses, priceFeedAddresses, owner);
        //engine.transferOwnership(address(mockDsce));
        // Arrange - User
        vm.startPrank(USER);
        ERC20Mock(weth).approve(address(mockDsce), amountCollateral);
        vm.expectRevert(DSCEngine.DSCEngine__MintFailed.selector);
        mockDsce.depositCollateralAndMintDsc(weth, amountCollateral, amountToMint);
        vm.stopPrank();
    }   

`

`
  function testRevertsIfMintAmountIsZero() public {
        vm.startPrank(USER);
        ERC20Mock(weth).approve(address(engine), amountCollateral);
        engine.depositCollateralAndMintDsc(weth, amountCollateral, amountToMint);
        vm.expectRevert(DSCEngine.DSCEngine_NeedsMoreThanZero.selector);
        engine.mintDsc(1);
        vm.stopPrank();
    }

`

The error message is below.

`
[FAIL. Reason: DSCEngine__BreaksHealthFactor(142857142857142857 [1.428e17])] testRevertsIfMintAmountIsZero() (gas: 171499)
Traces:
  [9042416] DSCEngineTest::setUp()
    ├─ [4037904] → new DeployDSC@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← 20105 bytes of code
    ├─ [4777813] DeployDSC::run()
    │   ├─ [3009444] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← ()
    │   │   ├─ [372455] → new MockV3Aggregator@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← 1417 bytes of code
    │   │   ├─ [449386] → new ERC20Mock@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   └─ ← 2502 bytes of code
    │   │   ├─ [372455] → new MockV3Aggregator@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   └─ ← 1417 bytes of code
    │   │   ├─ [449386] → new ERC20Mock@0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76
    │   │   │   └─ ← 2502 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← ()
    │   │   └─ ← 10470 bytes of code
    │   ├─ [912] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← MockV3Aggregator: [0x90193C961A926261B756D1E5bb255e67ff9498A1], MockV3Aggregator: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3], ERC20Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], ERC20Mock: [0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76], 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    │   ├─ [0] VM::startBroadcast(<pk>)
    │   │   └─ ← ()
    │   ├─ [591046] → new DecentralizedStableCoin@0x5FbDB2315678afecb367f032d93F642f64180aa3
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266)
    │   │   └─ ← 3261 bytes of code
    │   ├─ [937340] → new DSCEngine@0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    │   │   └─ ← 5057 bytes of code
    │   ├─ [2421] DecentralizedStableCoin::transferOwnership(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, newOwner: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   │   └─ ← ()
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← ()
    │   └─ ← DecentralizedStableCoin: [0x5FbDB2315678afecb367f032d93F642f64180aa3], DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], HelperConfig: [0x104fBc016F4bb334D775a19E8A6510109AC63E00]
    ├─ [912] HelperConfig::activeNetworkConfig() [staticcall]
    │   └─ ← MockV3Aggregator: [0x90193C961A926261B756D1E5bb255e67ff9498A1], MockV3Aggregator: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3], ERC20Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], ERC20Mock: [0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76], 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    ├─ [46789] ERC20Mock::mint(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 10000000000000000000 [1e19])
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], value: 10000000000000000000 [1e19])
    │   └─ ← ()
    └─ ← ()

  [171499] DSCEngineTest::testRevertsIfMintAmountIsZero()
    ├─ [0] VM::startPrank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D])
    │   └─ ← ()
    ├─ [24739] ERC20Mock::approve(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 10000000000000000000 [1e19])
    │   ├─ emit Approval(owner: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], spender: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], value: 10000000000000000000 [1e19])
    │   └─ ← true
    ├─ [126922] DSCEngine::depositCollateralAndMintDsc(ERC20Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], 10000000000000000000 [1e19], 100000000000000000000 [1e20])
    │   ├─ emit CollateralDeposited(user: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], token: ERC20Mock: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], amount: 10000000000000000000 [1e19])
    │   ├─ [30862] ERC20Mock::transferFrom(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 10000000000000000000 [1e19])
    │   │   ├─ emit Transfer(from: user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], to: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], value: 10000000000000000000 [1e19])
    │   │   └─ ← true
    │   ├─ [8993] MockV3Aggregator::latestRoundData() [staticcall]
    │   │   └─ ← 1, 700000000000 [7e11], 1, 1, 1
    │   ├─ [8993] MockV3Aggregator::latestRoundData() [staticcall]
    │   │   └─ ← 1, 100000000000 [1e11], 1, 1, 1
    │   └─ ← DSCEngine__BreaksHealthFactor(142857142857142857 [1.428e17])
    └─ ← DSCEngine__BreaksHealthFactor(142857142857142857 [1.428e17])
`

**Comment by EngrPips** [comment]
Hello @codexNature, The `healthFactor` function in your `DSCEngine` contract might currently lack this line of code `if (total_DSC_minted == 0) return type(uint256).max;` which makes a check to see if a user that has deposited some collateral minted any `DSC` yet, if yes that line of function would pass and the rest line of code in the  `healthFactor` would be executed but if no it would return `type(uint256).max` as the total `DSC` minted by the user which would prevent you from getting the current error you are getting, My implementation of the `healthFactor` function looks like below 

```javascript
function _health_factor(address _user) internal view returns (uint256 health_factor) {
        (uint256 total_DSC_minted, uint256 collateral_value_in_USD) = _get_account_information(_user);
        // console2.log("the total collateral value in USD is: ", collateral_value_in_USD);
        // console2.log("the total DSC minted is: ", total_DSC_minted);
        if (total_DSC_minted == 0) return type(uint256).max;
        uint256 collateral_amount_for_threshold =
            (collateral_value_in_USD * LIQUIDATION_THRESHOLD) / LIQUIDATION_THRESHOLD_DIVIDER;
        // console2.log("the collateral amount for threshold is: ", collateral_amount_for_threshold);
        health_factor = collateral_amount_for_threshold  / total_DSC_minted;
        // console2.log("the calculated health factor is: ", health_factor);
        return health_factor;
    }
```
